tensorboard/plugins/profile/end_to_end_util.py

```diff
@@ -0,0 +1,328 @@
+      model_fn_entire_duration = self._model_fn_end_time - self._model_fn_begin_time
+      return max(model_fn_entire_duration - self.SetupInfeedDuration(), 0)
+    return 0
+
+def _FillSessionBeginAndEndTimes(session_times, events):
+  for e in events:
+    if e.attribute == 'train_begin':
+      session_times['train'].session_begin_time = e.timestamp
+    if e.attribute == 'train_end':
+      session_times['train'].session_end_time = e.timestamp
+    if e.attribute == 'eval_begin':
+      session_times['eval'].session_begin_time = e.timestamp
+    if e.attribute == 'eval_end':
+      session_times['eval'].session_end_time = e.timestamp
+    if e.attribute == 'predict_begin':
+      session_times['predict'].session_begin_time = e.timestamp
+    if e.attribute == 'predict_end':
+      session_times['predict'].session_end_time = e.timestamp
+
+  for _, sess in six.iteritems(session_times):
+    if not sess.BeginAndEndTimesSet():
+      print(
+        'Either or both the begin and end time of session "%s" is not set'%sess.name
+      )
+    if sess.session_begin_time > sess.session_end_time:
+      print(
+        'Session begin time (%d) is bigger than session end time (%d) for session "%s"'%(
+          sess.session_begin_time, sess.session_end_time, sess.name)
+      )
+
+def _AssignEventsToSessions(session_times, events):
+  for e in events:
+    if e.attribute in set(['train_begin', 'train_end',
+                           'eval_begin', 'eval_end',
+                           'predict_begin', 'predict_end']):
+      continue
+    assigned = False
+    for _, sess in six.iteritems(session_times):
+      if sess.EventAssigned(e):
+        assigned = True
+        break
+    if not assigned:
+      print('Event "%s" is not assigned to any session'%e.attribute)
+
+  for _, sess in six.iteritems(session_times):
+    if not sess.AllTimesSet():
+      print('Not all the times in session "%s" are set'%sess.name)
+
+
+class EndToEndBreakDown(object):
+  def __init__(self, events):
+    session_times = {}
+    session_times['train'] = SessionTimeBreakDown('train')
+    session_times['eval'] = SessionTimeBreakDown('eval')
+    session_times['predict'] = SessionTimeBreakDown('predict')
+
+    _FillSessionBeginAndEndTimes(session_times, events)
+    _AssignEventsToSessions(session_times, events)
+
+    # All times are in milliseconds.
+    self._gap_train_begin_init_system_begin = 0
+    self._init_system_duration = 0
+    self._gap_init_system_end_model_fn_begin = 0
+    self._setup_infeed_duration = 0
+    self._model_fn_duration = 0
+    self._train_duration = 0
+    self._gap_eval_begin_last_session_end = 0
+    self._eval_duration = 0
+    self._gap_predict_begin_last_session_end = 0
+    self._predict_duration = 0
+    self._Compute(session_times['train'],
+                  session_times['eval'],
+                  session_times['predict'])
+
+  def _Compute(self, train_sess, eval_sess, predict_sess):
```

Comment:
There’s a lot of nontrivial logic here. Please test this module.

Reply:
A unit test for it (end_to_end_util_test.py) is added.